### ISSUE TITLE
Add support for dependency injection

### ISSUE THREAD

**Issue by gitfool** [xunit/xunit#1738]
We've discussed this previously in #687 but perhaps it could be reconsidered for 3.0? 😉 

I'll open with a comment I made at the time that's still relevant today:
> We like xUnit to the degree that it made sense to leverage its familiar framework and runners to also drive our integration tests. Both Devs & QA love the consistency. As with unit tests we can drive them from within Visual Studio, with Resharper, and the commandline. This has been invaluable, but is obviously where we heavily use DI.

I'd like support for DI to be officially baked in and based on past hacking that involves the following:

- Assembly fixtures
  - e.g. [kzu/xunit.assemblyfixture](https://github.com/kzu/xunit.assemblyfixture)
- Global fixture(s)
  - i.e. across all loaded assemblies (when multiple specified)
- Predefined lifetime scopes
  - per test (transient), class, collection, assembly, global (singleton)
  - ... as per fixtures but with first class support these could replace fixtures
- Container bootstrapping
  - xunit type registration; e.g. `ITestOutputHelper` per test
  - user type registration hooks

I'm most familiar with [Autofac](https://www.fuget.org/packages/Autofac), but this task would presumably have to be container agnostic, although [Microsoft.Extensions.DependencyInjection](https://www.fuget.org/packages/Microsoft.Extensions.DependencyInjection) might be an option now.

**Comment by tai-yi** [comment]
As a dotnetcore developer, I prefer  `Microsoft.Extensions.DependencyInjection`.

**Comment by bartelink** [comment]
I know I could just thumbs down, but as a dotnetcore developer, I do not prefer `Microsoft.Extensions.DependencyInjection` and would hate to see a dependency on it. (I trust others, including Brad, will have similar opinions)

**Comment by tai-yi** [comment]
Developers have different love for DI container.
It's reasonable to reduce dependency, so the better way is setup an interface such as below code.
Then we can write different XunitStartupPlugins.

``` c#
    public interface IXunitStartup<T, TScope>
        where TScope : IDisposable
    {
        //Root DI container
        T ServiceProvider { get; }

        TScope CreateScope(T sp);

        TServive GetService<TServive>();

        TServive GetService<TServive>(TScope scope);

        IEnumerable<TService> GetServices<TService>();

        IEnumerable<TService> GetServices<TService>(TScope scope);

    }
    public class XunitStartupAttribute : Attribute { }

    [XunitStartup]
    public class MsDiXunitStartup : IXunitStartup<IServiceProvider, IServiceScope>
    {

        private IServiceProvider _rootSp;

        public IServiceProvider ServiceProvider
        {
            get
            {
                if (_rootSp == null)
                {
                    var services = new ServiceCollection();
                    //setup my services
                    _rootSp = services.BuildServiceProvider(true);
                }
                return _rootSp;
            }
        }

        public TServive GetService<TServive>(IServiceScope scope)
        {
            return scope.ServiceProvider.GetRequiredService<TServive>();
        }

        public IServiceScope CreateScope(IServiceProvider sp)
        {
            return sp.CreateScope();
        }

        public TServive GetService<TServive>()
        {
            return ServiceProvider.GetRequiredService<TServive>();
        }

        public IEnumerable<TService> GetServices<TService>()
        {
            return ServiceProvider.GetServices<TService>();
        }

        public IEnumerable<TService> GetServices<TService>(IServiceScope scope)
        {
            return scope.ServiceProvider.GetServices<TService>();
        }
    }
```

**Comment by tullo-x86** [comment]
@tai-yi:

The reason this hasn't happened yet is because it's easy for such "conforming containers" [to turn into an anti-pattern](https://blog.ploeh.dk/2014/05/19/conforming-container/#68e62f6ecc7643c18c0616c2dc405a94). There [are ways to make a framework DI-friendly](https://blog.ploeh.dk/2014/05/19/di-friendly-framework/) without that, however...

**Comment by he-dev** [comment]
@tullo-x86 true, however, this is rather a workaround than a real sollution. Being able to plung-in **any** DI would be a great improvement.

@tai-yi well, I prefer `Autofac`... so a general interface would be a much better choice.

**Comment by pengweiqhca** [comment]
https://www.nuget.org/packages/Xunit.DependencyInjection/

**Comment by gitfool** [comment]
Sorry to resurrect an old wound but I just had an epiphany while pondering getting dotnet core configuration into unit / integration tests using xunit. If we can require dotnet core, then for configuration, logging and dependency injection, connecting the dots, what we want is for tests to bootstrap and run via the new [generic host](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-3.1). That would be perfect!

@pengweiqhca at a glance, is it fair to say that's the approach for your solution above?

**Comment by pengweiqhca** [comment]
> Sorry to resurrect an old wound but I just had an epiphany while pondering getting dotnet core configuration into unit / integration tests using xunit. If we can require dotnet core, then for configuration, logging and dependency injection, connecting the dots, what we want is for tests to bootstrap and run via the new [generic host](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-3.1). That would be perfect!
> 
> @pengweiqhca at a glance, is it fair to say that's the approach for your solution above?

Yes, but only partially

**Comment by MortenChristiansen** [comment]
I would really love to see support for property injection. I often have test base classes with one or more constructor arguments. Adding a new argument requires me to update all the implementing test classes which can be a pain. With the new 'required' keyword these properties could be available in the constructor.

**Comment by bradwilson** [comment]
I am going to close this issue.

I don't believe that it's beneficial for us to "pick one" dependency injection system as the official system. For v3, we've added to ability to shim in very early--before discovery or execution begins--by way of implementing `ITestPipelineStartup`. I would expect adapters for common DI patterns/frameworks would use this to ensure that any kind of startup code (including setting up DI containers) would be run.

If there are specific technical issues around this functionality, please open up new issues. This single wide-ranging issue is no longer useful for that purpose.

Thanks!
